### PR TITLE
Optional game rule for defender retreat

### PR DESCRIPTION
**Pull request by kevinkunker**
## Notes to Reviewer
On this fork I added a new unit attachment called "canDefensiveRetreat", which when set to true, adds an option for that unit to retreat from battle when defending.

### CODE REVIEW

**Comment by Cernelius** [comment]
Very interesting and valuable new feature, but please make sure that, if the property "Retreating Units Remain In Place" is true, defending units too retreat to the same territory as the battle territory (sort of like when submerging submarines).

**Comment by kevinkunker** [comment]
@Cernelius Thank you! Yes the code checks for that property, as well as things like enemy units blocking retreat, if the defender can move through canals, etc.

**Comment by Cernelius** [comment]
Why not offering defenders the option to retreat before everything else, meaning before the first round of combat (so possibly retreating without fighting at all)? Since the attackers are usually attacking willingly, it is logical not to ask them to retreat before at least one round of combat is done (because it would be just weird that you move into for an invasion and retreat before firing a shot), but I think that the defenders can decide they want none of it right away before the battle even starts.

If not having defensive retreat before everything else, how about having an additional setting for deciding whether defensive retreat is at the start or at the end of the combat round? I think that would be very good.

(The behaviour of retreating before the battle even starts would be absolutely realistic in the case of battles between surface ships of similar maximum speed or cavalry clashes. Other cases are disputable.)

**Comment by WCSumpton** [comment]
@kevinkunker

This does not follow the "Defender Fight or Retreat (Land Combats Only)" rule as outlined in North Africa rulebook page 20.
Defender must declare, at the beginning of the combat round, whether to fight or retreat. If the defender choses to retreat, those retreating units will still be fired upon by the attacker, but none of retreating units will fire back, and any such surviving units must leave the territory.
**Note:** Because they fire _before_ combat begins, units performing Air Defense may fire even if retreating.
Since TripleA does not have a **dedicated** Air Defense phase/step, how would this be handled?
Should there be two defender retreat options for land combat, one for "Fight or Retreat", the overriding option, and another for "Retreat After Combat".
Should defender navel retreat be separated from land combat, and what are the options for defender navel retreats since North Africa does not allow defensive navel retreats?
If defending land units are allowed to retreat into the same territory, "Retreating Units Remain In Place", what happens to the control of the "retreated" territory? Does the defender retain control or does the act of retreating give control of the territory to the attacker?
Can an attacker still preform an Amphibious Assault from a sea zone in which the defender has retreated surface warships?
TripleA does need a Defender Retreat. This PR, IMO, needs to be a little more robust so that it includes the "official" rules as describe in North Africa and more what-ifs testing on the effects on present gameplay.
There are many "tack-on" rules and options in TripleA which either violate "official" rules, have bad gameplay outcomes, etc... This outstanding work done here, IMO, should not be added in an incomplete state.

The above are just my thoughts. I still think this is a fantastic beginning to solving a problem many believe has persisted to long.

Cheers...

**Comment by Cernelius** [comment]
I assumed this feature was not based on any particular existing ruleset. Is it so or is this meant to support the North Africa game?

In the former case, how about coding it with enough properties as to ensure the game-maker can use the feature either as intended by @kevinkunker or so to enforce the relative rules of the North Africa game?

In the latter case, I suggest having a property for switching on/off the rule of getting fired on by the attacker upon retreating (meaning that, if this property is true meaning off, you can retreat at the start of the combat round without taking any losses on that round - or at all if you decide to retreat at the start of the first round of combat).

**Comment by WCSumpton** [comment]
@Cernelius 

> I assumed this feature was not based on any particular existing ruleset.

This is my thought also. IMO there should be two options for land units "Defender Can Retreat Land Units After Combat" which would be this PR, and another option called "Defender Can Fight or Retreat Land Units" which follow the rules in North Africa. I also think defender retreating of navel units should be a separate option.
@kevinkunker does not have to change this PR, only restate the option, and someone else can code the North Africa rule later.

As to my other questions. Who owns the territory if the defender chooses to retreat with "Retreating Units Remain In Place"? And can an attacker still preform an Amphibious Assault from a sea zone where the defender has retreated surface warships?

Cheers...

**Comment by kevinkunker** [comment]
@WCSumpton If the defender chooses to retreat with "Retreating Units Remain In Place", then the attacker takes over the territory and the attacker is marked as winning in the battle statistics. A defender can't retreat into a territory where enemy units are present, so the second scenario should never happen. @Cernelius Yeah this feature was not based on any particular rule set, I was just thinking of generalizing the sub retreat for all units. If it ends up being useful for maps like North Africa then that will be great but wasn't what I was originally thinking of (maybe I could try implementing that next as a separate pull request?).

Also after more playtesting, I discovered a bug where factories and other 0 move units could retreat, so I am making a commit to fix that now. Can't imagine the building growing legs and running away...

**Comment by Cernelius** [comment]
> @Cernelius Yeah this feature was not based on any particular rule set, I was just thinking of generalizing the sub retreat for all units.

Good idea. I think that you were mainly thinking about V1.0 (meaning Classic 2nd Edition (non-3rd Edition nor Iron Blitz)). However, in this case I'll advocate even more strongly (as part of this PR) for a property allowing setting the retreat at the start of the combat round (like the subs of V3+) rather than at the end of it (like the subs of V2-).

**Comment by Cernelius** [comment]
> Also after more playtesting, I discovered a bug where factories and other 0 move units could retreat, so I am making a commit to fix that now. Can't imagine the building growing legs and running away...

I suggest doing this with a unit option which determines whether or not that unit retreats (both offensively and defensively). In turn, such unit option, when not set, should default to "true" if the unit has movement 1 or more and "false" if the unit has movement 0.

This way, you would also provide a fix for this other issue (killing two birds with a stone):
https://github.com/triplea-game/triplea/issues/7167

You can call this option "cannotRetreat" or "emplacement" or "territorial" or "fixed".

**Comment by WCSumpton** [comment]
@kevinkunker 

> Also after more playtesting, I discovered a bug where factories and other 0 move units could retreat, so I am making a commit to fix that now. Can't imagine the building growing legs and running away...

This is the same problem that plagued the 'Active Unit Scroller' (#14058). There are two conditions, that I am aware of, that will allow a 0 movement unit to move. Can the unit be transported, have a 'transportCost' greater then 0 and a unit that can transport it is present. Can the unit receive movement from another unit, a unit has 'givesMovement' and is present.
A PR was merged (#14060), maybe this could help here.

Cheers...

**Review by frigoref — CHANGES_REQUESTED** [review]
@kevinkunker Before I provide a complete review, please write at least 2 test methods (one positive and one negative) into a corresponding test class to verify the new functionality. It is advisable to write also test methods for each side effect you have encountered. 
If you need help with that, please let us know.
In any case, we aligned to have more test cases in the future to better stabilize the engine.

**Comment by DanVanAtta** [comment]
Thank you Kevin for proposing the update and your efforts. They are really appreciated and welcome. Generally agree and would second the various discussion points. 

FWIW, a new property is a pretty big deal. Mainly from the perspective that game mechanics interact, and so future mechanics will have to take into account the prior. To add to the long lasting consequences, XML properties have indefinite support.

With all that said, my initial thought is we would want more flexibility. Namely:

- instead of true/false, how about setting the property as which round a defender could retreat? Eg: -1 is can't, 0 is start of first round, 1 is start of second, etc..

- perhaps flags (additional XML subelements) for whether infrastructure or zero move can retreat

- flags for if retreat should be whole army, unit types, or individual units 

- flag for if retreat can be sane territory, or an eligible move territory

There is quite a bit to consider overall. Let us know your thoughts pls. I would suggest opening a forum post to open these game design questions to the larger community, gather info on how the feature would want to be used by others and to talk through the implications of the new mechanic.

I agree that there should be a testing strategy. Largely automated and some manual.  I'm curious for your thoughts there as well

**Comment by Cernelius** [comment]
> * perhaps flags (additional XML subelements) for whether infrastructure or zero move can retreat

"Infrastructure" is just a case of bad or legacy naming. It has nothing to do with movement so should be unrelated to retreat as well.

(Actually, "infrastructure" is just wrong a name anyway even for every basic game from V1 to V5 (where infrastructures are always and only AA Guns and Factories): an AA Gun is obviously not an infrastructure, but a Factory is also not an infrastructure, though in this case at least you can say that the factory unit itself necessarily comprises the needed infrastructures (railways, roads, water system...) for the factories, barracks and so on.)

**Comment by Cernelius** [comment]
https://forums.triplea-game.org/topic/1644/a-better-word-than-infrastructure?_=1782324993208

File: game-app/game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java [hunk 1]
```diff
@@ -866,6 +867,63 @@ public Collection<Territory> getAttackerRetreatTerritories() {
...
+    // the air unit may have come from a conquered or enemy territory, don't allow retreating
+    final Predicate<Territory> conqueredOrEnemy =
+        Matches.isTerritoryEnemyAndNotUnownedWaterOrImpassableOrRestricted(defender)
+            .or(Matches.territoryIsWater().and(Matches.territoryWasFoughtOver(battleTracker)));
+    possible.removeAll(CollectionUtils.getMatches(possible, conqueredOrEnemy));
+
+    if (defendingUnits.stream().anyMatch(Matches.unitIsLand()) && !battleSite.isWater()) {
+      possible = CollectionUtils.getMatches(possible, Matches.territoryIsLand());
+    }
+    if (defendingUnits.stream().anyMatch(Matches.unitIsSea())) {
```
Comment (DanVanAtta) [inline game-app/game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java:920]: I believe there is a bug here. If we have a carrier with fighters, and the carrier is sunk - will the defending units still contain the carrier? If no, then this logic is skipped. That means any fighters on said carriers, will have different movement options depending on whether the carrier is present or not.

Overall:
1.A) I do think we should pause and design how the defensive retreat mechanic should work in detail. I was wondering a lot if defensive retreat could even be used to "advance" units, and how the rules for where to retreat would work.
1.B) There is an open issue that sub submerge should not be all-or-none, but an arbitrary number. Retreat seems like it should be similar.
1.C) Gameplay, could retreat be easily abused? Like, if playing on LL, the max number of hits is bounded, couldn't a person take advantage of that and land some planes in a territory that could otherwise be captured, but knowingly retreat them after a first round of combat? 
2) I'm pretty sure we should not use properties as heavily as we do. I wonder when/how to rework that and am bit hesitant to further add more properties. That consideration more reflects updates for Constants and Properties, and less here. But it's an issue I'm wondering about a lot.
3) The mechanics for retreat is something to be tested. Which territories & which units are key things that could be validated.

Comment (kevinkunker) [reply game-app/game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java:920]: The way the engine works is that air units don't retreat with other land/sea units, they always remain in place when retreating, even with other friendly land/sea units that are able to retreat (and that is not something I added, that is just the default behavior). Also the piece of code which is cut off below is only for determining if sea units can pass through a canal when retreating, it doesn't affect the air units (which will be left in place).

1.A) defensive retreat cannot be used to advance units in any manner, it is explicitly not allowed for units to retreat to an enemy territory or where enemy units are present
1.B) (EDIT: actually looking back over it, it may be possible?)
1.C) The planes would still get hit (assuming retreat after combat)? the territory would be captured by the enemy anyway and the planes would be lost? I genuinely don't see how this could be advantageous in any way

EDIT: Removed stuff saying I couldn't get the int property working, it is back to working again

Personally I don't really like the idea of having defenders retreating before combat because it is anticlimactic and leads to wild goose chases, but it is there if someone wants it

File: gradle.properties [hunk 2]
```diff
@@ -9,3 +9,5 @@ org.gradle.jvmargs=--add-exports jdk.compiler/com.sun.tools.javac.api=ALL-UNNAME
   --add-exports jdk.compiler/com.sun.tools.javac.parser=ALL-UNNAMED \
   --add-exports jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED \
   --add-exports jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED
+
+org.gradle.jvmargs=-Dfile.encoding=UTF-8
```
Comment (frigoref) [inline gradle.properties]: Guess this has nothing specific to do with this issue and should be carved out if needed.

File: game-app/game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java (hunk 1 above)
Comment (DanVanAtta) [reply game-app/game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java:920]: > The way the engine works is that air units don't retreat with other land/sea units, they always remain in place when retreating

For offensive, but why should defensive retreat have to work in the same way? If so, retreat + planes is a way to suicide the planes.

Thinking about retreat options... it really comes down to a unit ability. It's not a battle feature, but a unit ability.

We have to weigh how to be pragmatic here.

On the one hand, the way battles are done needs rework. A lot of TripleA XML and engine logic needs rework so that more of the config is driven by the XML rather than assumed the engine.

For example, there should be a "round 0" in combat. Each unit should have a "start-fire round", and "end-fire round" setting. Bships doing land battles would have a "start-fire" and "end-fire" of round 0 for classic rules and round 1 for revised rules (which allows units to return fires). I suspect almost all properties need a rework like this. These types of combat rules should be attached to a new concept of "unit category" and then each unit assigned a unit-category, and/or given its own combat rules individually.

The pragmatism is how much value we get out of a quicker win, but one in which we will also need to forever support new XML properties. 

> 1.A) defensive retreat cannot be used to advance units in any manner, it is explicitly not allowed for units to retreat to an enemy territory or where enemy units are present

I'm considering things like strafing combats, where planes attack land units. Also another scenario is where there adjacent stacks, retreating into the adjacent stack is effectively a way to add a movement point.

> Personally I don't really like the idea of having defenders retreating before combat because it is anticlimactic and leads to wild goose chases, but it is there if someone wants it

I have always thought Civil War era maps would benefit from defensive retreat. Having the ability be per-unit makes the feature a bit more interesting in a number of ways. From a game design perspective, it makes more sense as well I believe. I'm still thinking a lot about the implications of such a feature. I'll emphasize this kind of update is one of the least trivial updates you could make! A very major UI feature would require less scrutiny than a XML property that adds a new game mechanic. I believe we need to be sure we think this through well and add the mechanic the right way. Otherwise we'll have "2 maps" that have special rules that we'll be forced to work around and continue supporting for a long time. There's a number of such rules and they are a PITA code-wise and several times it would have been easier to drop those features (and those features existed have created blockers for others). So.. we need to proceed with caution here.

File: game-app/game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/DefensiveGeneralRetreat.java [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+                      < unit.getUnitAttachment().getDefensiveRetreatRound());
+
+      // Remove units that can't retreat due to territory effect
+      TerritoryEffectAttachment territoryEffect =
+          (TerritoryEffectAttachment) retreatTo.getAttachment(TERRITORYEFFECT_ATTACHMENT_NAME);
+      if (territoryEffect != null)
+        retreater
+            .getRetreatUnits()
+            .removeIf(unit -> territoryEffect.getUnitsNotAllowed().contains(unit));
+
```
Comment (VictoryFirst2) [inline game-app/game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/DefensiveGeneralRetreat.java]: Please add a check that removes units that can't retreat to territories due to the relationship with the owner.

File: game-app/game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java [hunk 4]
```diff
@@ -866,6 +871,65 @@ public Collection<Territory> getAttackerRetreatTerritories() {
...
+    Collection<Territory> possible =
+        CollectionUtils.getMatches(
+            getGameData().getMap().getNeighbors(getBattleSite()),
+            Matches.territoryHasUnitsThatMatch(enemyUnitsThatPreventRetreat).negate());
+
+    // Remove territories we can't retreat to due to territory effects
+    final Predicate<Territory> unitsNotAllowed =
+        Matches.territoryEffectsAllowUnits(defendingUnits).negate();
+    possible.removeAll(CollectionUtils.getMatches(possible, unitsNotAllowed));
+
```
Comment (VictoryFirst2) [inline game-app/game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java:906]: Please add a check that removes units that cannot retreat to a territory due to the relationship with the owner.

File: game-app/game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/BattleStep.java [hunk 5]
```diff
@@ -45,6 +47,7 @@
 public interface BattleStep extends IExecutable {
 
   enum Order {
+    DEFENDER_FIGHT_OR_RETREAT,
```
Comment (VictoryFirst2) [inline game-app/game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/BattleStep.java:50]: This causes NoSuchElementExceptions when you have to deal with AA, naval bombard and/or first strike fire/casualties. I would put it directly above "GENERAL_OFFENSIVE", also in line with the rules for North Africa.

File: game-app/game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java [hunk 6]
```diff
@@ -866,6 +871,65 @@ public Collection<Territory> getAttackerRetreatTerritories() {
     return possible;
   }
 
+  @Override
+  public Collection<Territory> getDefenderRetreatTerritories() {
+
+    // If defender is all planes, just return collection of current territory
```
Comment (VictoryFirst2) [inline game-app/game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java:877]: You can do it this way, but I'd instead make sure that air units retreat along with the other units, and not hover over the territory until the it's the defenders turn. This will prevent NCMs into the territory as well as income collected if "Contested Territories Produce No Income" is true. You can always implement both and use a game property to govern this.

**Review by VictoryFirst2 — CHANGES_REQUESTED** [review]
Right now it is looking very promising. I'd love to see this implemented. I think it needs thorough testing of edge cases to make sure it works bug-free. Consider some regression tests for these.
Examples:
- suicide units
- naval bombard with/without casualties return fire
- submarine first strike and submerging
- retreating into contested territories
- ...

We need to discuss the exact details of the above tests.

I'd also suggest to construct a detailed overview of the new functionalities and how certain situations are handled, so that we can discuss these as well. If you can put something like that in the PR description, that would be great.

File: game-app/game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java (hunk 4 above)
Comment (kevinkunker) [reply game-app/game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java:906]: Thanks for the feedback, just added this

Comment (VictoryFirst2) [reply game-app/game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java:906]: Great. Just realized I misworded my feedback but fortunately you understood 😁 Meant to say "...that removes territories which units cannot retreat to..."

**Comment by Cernelius** [comment]
Quick question (since this started as a custom retreat feature and became North Africa's), is it still true that with "Retreating Units Remain In Place" defensive units stay where they are? (Hope so.)

By the way, I think that with defensive suicide units it would be especially valuable to allow to choose how many of them you can retreat (so to avoid them suiciding on an overkill).

As of now, suicide units are almost unusable in defence because they can be all wasted by the attacker sending a single unit in the zone to trigger them all suiciding to defend.

**Comment by kevinkunker** [comment]
@Cernelius Yes "Retreating Units Remain In Place" still works, all of that got carried over. That is interesting with the defending suicide units, it seems like maybe there should be some restriction where the number of times they fire/detonate shouldn't be more than the number of the attacker's hitpoints.

**Comment by VictoryFirst2** [comment]
@kevinkunker Thanks for the new commits, this is looking already very good in my opinion. It now just needs thorough testing of edge cases. For example:
- a unit is hit by a bombard but is still able to return fire because of the game properties. What happens if the defender decides to retreat? If that unit was the only unit defender, is the option to retreat still offered?
- A naval unit is hit by a first strike. If it's the only naval unit that can retreat, is there an option to retreat? Does it not fire back in that case?
- Can defenders retreat from a strategic bombing raid, or only from a normal combat? 
- If there is no retreat territory possible, is the option to retreat not offered?
- If there are only suicide units on the defender's side, is the option to retreat offered?
- When units with zero defense value are defending (e.g. transports) , can they defensively retreat? Or are they destroyed before they get that option?
- What happens if a transport can defensively retreat, but not the land units being carried by it? Do the land units retreat along with the transport?


These are just some things I was able to come up with at the moment. I don't think every situation needs a test case, some can be checked with the UI as well. I'd be happy to help with testing these situations.

The defensive retreat rules are now almost fully implemented. There are just two things that would be nice to have added, but are in my opinion not a necessity for merging - the PR in its current state plus a verification of at least the above situations is already quite good - so they are more like feature requests.
- Add a unit option "isForcedToDefensivelyRetreat", or "mustDefensiveRetreat", and then a round number specifying during which round this unit must retreat from battle. This would be handy for the scout cars
- Allow the defender to choose which units retreat to which territory, instead of selecting one territory and let all units retreat only to that territory. This may be hard to implement.

**Comment by DanVanAtta** [comment]
@kevinkunker / @VictoryFirst2 To avoid boiling the issue, I propose we make this feature multi-phase. In part to buy time to hash things out but also allow forward progress.

To that, I recommend these actions:
- polish this branch:  do a bug-find & fix pass, resolve nitpicks, tighten up commentary, etc
- put the new behavior in this branch  behind a feature flag, then we can merge this. The point of the feature flag is to help clarify that this is still an in-progress feature (which alleviates my concern of perma-supporting legacy XML specs) 
- open an issue or a discussion thread to track where we are and next steps
  - Victory mentioned some important notes: https://github.com/triplea-game/triplea/pull/14649#issuecomment-5432099036
  - this new issue will help make sure we don't lose any important comments from this PR and overall help us stay organized. 

I'll have a robot do an automated review, fixing those would IMO polish up the branch reasonably enough.
